authcrypt/main.cpp

```diff
@@ -53,132 +39,177 @@ static const unsigned char secret_key[16] = {
+         *
+         * First generate a random 8-byte nonce.
+         * Put it directly in the output buffer as the recipient will need it.
+         *
+         * Warning: you must never re-use the same (key, nonce) pair. One of
+         * the best ways to ensure this to use a counter for the nonce.
+         * However, this means you should save the counter accross rebots, if
+         * the key is a long-term one. The alternative we choose here is to
+         * generate the nonce randomly. However it only works if you have a
+         * good source of randomness.
+         */
+        const size_t nonce_len = 8;
+        mbedtls_ctr_drbg_random(&drbg, ciphertext, nonce_len);
+
+        size_t ciphertext_len = 0;
+        /* Go for a conservative 16-byte (128-bit) tag
+         * and append it to the ciphertext */
+        const size_t tag_len = 16;
+        ret = mbedtls_cipher_auth_encrypt(&cipher, ciphertext, nonce_len,
+                            reinterpret_cast<const unsigned char *>(metadata),
+                            sizeof(metadata),
+                            reinterpret_cast<const unsigned char *>(message),
+                            sizeof(message),
+                            ciphertext + nonce_len, &ciphertext_len,
+                            ciphertext + nonce_len + sizeof(message),
+                            tag_len);
+        if (ret != 0) {
+            mbedtls_printf("mbedtls_cipher_auth_encrypt() returned -0x%04X\r\n",
+                           -ret);
+            return 1;
+        }
+        ciphertext_len += nonce_len + tag_len;
+
+        /*
+         * The following information should now be transmitted:
+         * - First ciphertext_len bytes of ciphertext buffer
+         * - Metadata if not already transmitted elsewhere
+         */
+        print_hex("ciphertext", ciphertext, ciphertext_len);
+
+        /*
+         * Decrypt-authenticate
+         */
+        size_t decrypted_len = 0;
+
+        ret = mbedtls_cipher_setkey(&cipher, secret_key,
+                                    8 * sizeof(secret_key), MBEDTLS_DECRYPT);
+        if (ret != 0) {
+            mbedtls_printf("mbedtls_cipher_setkey() returned -0x%04X\r\n",
+                           -ret);
+            return 1;
+        }
+
+        ret = mbedtls_cipher_auth_decrypt(&cipher, ciphertext, nonce_len,
+                        reinterpret_cast<const unsigned char *>(metadata),
+                        sizeof(metadata), ciphertext + nonce_len,
+                        ciphertext_len - nonce_len - tag_len, decrypted,
+                        &decrypted_len, ciphertext + ciphertext_len - tag_len,
+                        tag_len);
+        /* Checking the return code is CRITICAL for security here */
+        if (ret == MBEDTLS_ERR_CIPHER_AUTH_FAILED) {
+            mbedtls_printf("Something bad is happening! Data is not "
+                           "authentic!\r\n");
+            return 1;
```

Comment:
Why are we changing return from ```ret``` to ```0```? At https://github.com/ARMmbed/mbed-os-example-tls/pull/95/files#diff-64da713f1246019fb29c1e604285e6e0R211 the return code is printed again and will always have ```1```.
Applies to all places returning 1.